runtime/lua/vim/diagnostic.lua

```diff
@@ -2,22 +2,95 @@ local api, if_nil = vim.api, vim.F.if_nil
+--- @alias Diagnostic vim.Diagnostic
+--- @class vim.Diagnostic
+--- @field bufnr? integer
+--- @field lnum integer 0-indexed
+--- @field end_lnum? integer 0-indexed
+--- @field col integer 0-indexed
+--- @field end_col? integer 0-indexed
+--- @field severity? vim.diagnostic.Severity
+--- @field message string
+--- @field source? string
+--- @field code? string
+--- @field _tags? { deprecated: boolean, unnecessary: boolean}
+--- @field user_data? any arbitrary data plugins can add
+--- @field namespace? integer
+
+--- @class vim.diagnostic.Opts
+--- @field float? boolean|vim.diagnostic.Opts.Float
+--- @field update_in_insert? boolean
+--- @field underline? boolean|vim.diagnostic.Opts.Underline
+--- @field virtual_text? boolean|vim.diagnostic.Opts.VirtualText
+--- @field signs? boolean|vim.diagnostic.Opts.Signs
+--- @field severity_sort? boolean|{reverse?:boolean}
+
+--- @class vim.diagnostic.OptsResolved
+--- @field float vim.diagnostic.Opts.Float
+--- @field update_in_insert boolean
+--- @field underline vim.diagnostic.Opts.Underline
+--- @field virtual_text vim.diagnostic.Opts.VirtualText
+--- @field signs vim.diagnostic.Opts.Signs
+--- @field severity_sort {reverse?:boolean}
+
+--- @class vim.diagnostic.Opts.Float
+--- @field bufnr? integer
+--- @field namespace? integer
+--- @field scope? 'line'|'buffer'|'cursor'|'c'|'l'|'b'
+--- @field pos? integer|{[1]:integer,[2]:integer}
+--- @field severity_sort? boolean|{reverse?:boolean}
+--- @field severity? vim.diagnostic.Severity
+--- @field header? string|{[1]:string,[2]:any}
+--- @field source? boolean|string
+--- @field format? fun(diagnostic:vim.Diagnostic): string
+--- @field prefix? string|table|(fun(diagnostic:vim.Diagnostic,i:integer,total:integer): string, string)
+--- @field suffix? string|table|(fun(diagnostic:vim.Diagnostic,i:integer,total:integer): string, string)
+--- @field focus_id? string
+
+--- @class vim.diagnostic.Opts.Underline
+--- @field severity? vim.diagnostic.Severity
+
+--- @class vim.diagnostic.Opts.VirtualText
+--- @field severity? vim.diagnostic.Severity
+--- @field source? boolean|string
+--- @field prefix? string|function
+--- @field suffix? string|function
+--- @field spacing? integer
+--- @field format? function
+--- @field hl_mode? 'replace'|'combine'|'blend'
+--- @field virt_text? {[1]:string,[2]:any}[]
+--- @field virt_text_pos? 'eol'|'overlay'|'right_align'|'inline'
+--- @field virt_text_win_col? integer
+--- @field virt_text_hide? boolean
+
+--- @class vim.diagnostic.Opts.Signs
+--- @field severity? vim.diagnostic.Severity
+--- @field priority? integer
+--- @field text? table<vim.diagnostic.Severity,string>
+--- @field numhl? table<vim.diagnostic.Severity,string>
+--- @field linehl? table<vim.diagnostic.Severity,string>
+--- @field texthl? table<vim.diagnostic.Severity,string>
+
+--- @alias DiagnosticSeverity vim.diagnostic.Severity
+--- @enum vim.diagnostic.Severity
 M.severity = {
   ERROR = 1,
   WARN = 2,
   INFO = 3,
   HINT = 4,
+  [1] = 'ERROR',
```

Comment:
Why not use `vin.tbl_add_reverse_lookup`?

Reply:
Because that doesn't add any typing and we plan to deprecate that function.